### ISSUE TITLE
[speckit.plan] Audit Pipelines Page for UI Consistency, Quality & UX in Project Solune

### ISSUE THREAD

**Issue by Boykai** [Boykai/github-workflows#3089]
> **Parent Issue:** #3086 — Audit Pipelines Page for UI Consistency, Quality & UX in Project Solune

## 🤖 Agent Task: `speckit.plan`

Create a detailed implementation plan. Break down the specification into actionable steps, identify dependencies, and define the order of execution.

---

## Overview
Conduct a comprehensive audit of every element and component on the Pipelines page within Project Solune. The audit should evaluate visual cohesion with the rest of the application, identify any errors or bugs, assess adherence to modern development best practices, and ensure the page delivers the best possible user experience.

## Technical Details
- Review all UI components on the Pipelines page against the established design system/style guide used across the rest of Project Solune
- Inspect for visual inconsistencies such as mismatched typography, color usage, spacing, iconography, and layout patterns
- Check for functional bugs, broken interactions, edge case handling (e.g. empty states, loading states, error states), and console errors
- Evaluate component architecture for adherence to modern best practices (e.g. reusability, separation of concerns, accessibility standards such as WCAG 2.1)
- Assess UX quality including clarity of information hierarchy, responsiveness, interaction feedback, and overall usability
- Document all findings with severity levels (Critical / High / Medium / Low) and proposed remediation steps

## Acceptance Criteria
- [ ] All visual inconsistencies between the Pipelines page and the rest of the application are identified and documented
- [ ] All functional bugs, broken interactions, and edge case gaps (empty, loading, error states) are catalogued with reproduction steps
- [ ] Code quality and best practice violations are flagged with specific file/component references and recommended fixes
- [ ] Accessibility issues are identified and mapped to relevant WCAG 2.1 guidelines
- [ ] A prioritized audit report is delivered, covering all findings with severity ratings and actionable remediation recommendations

---
*Sub-issue created for agent `speckit.plan` — see parent issue #3086 for full context*

**Comment by Boykai** [comment]
**`.github/agents/copilot-instructions.md`** (generated by `speckit.plan`)

---

# GitHub Workflows Chat — Development Guidelines

Last updated: 2026-03-10

> **Important:** Always use search tools (e.g., Context7 MCP, Microsoft Docs MCP, or web search) to look up the most up-to-date documentation when working with any and all libraries, frameworks, and APIs. Never rely solely on training data — verify current syntax, options, and best practices from official sources before writing or modifying code.

## Current Stack

### Backend

- **Runtime floor:** Python `>=3.12` (`backend/pyproject.toml`); primary dev/runtime target is Python 3.13 (`ruff` target `py313`, `pyright` `pythonVersion = "3.13"`, Docker image `python:3.13-slim`)
- **Framework:** FastAPI `>=0.135.0`, Uvicorn `>=0.41.0`
- **GitHub integration:** `githubkit>=0.14.6`, `httpx>=0.28.0`
- **Validation / config:** `pydantic>=2.12.0`, `pydantic-settings>=2.13.0`
- **Storage:** SQLite via `aiosqlite>=0.22.0` (WAL mode, single persistent connection, migrations run on startup)
- **AI providers:** `github-copilot-sdk>=0.1.30` (default), `openai>=2.26.0`, `azure-ai-inference>=1.0.0b9` (optional fallbacks)
- **Security / crypto:** `cryptography>=44.0.0` (Fernet token-at-rest encryption)
- **Rate limiting:** `slowapi>=0.1.9`
- **Utilities:** `tenacity>=9.1.0`, `websockets>=16.0`, `python-multipart>=0.0.22`, `pyyaml>=6.0.3`
- **Dev tools:** `ruff>=0.15.0`, `pyright>=1.1.408`, `pytest>=9.0.0`, `pytest-asyncio>=1.3.0`, `pytest-cov>=7.0.0`

### Frontend

- **Node / build:** Node 22 for local dev and Docker; Vite 7.3 (`frontend/vite.config.ts` is the single source of truth)
- **Framework:** React 19.2, react-router-dom v7
- **Language:** TypeScript ~5.9 (strict mode, `@/` alias → `frontend/src`)
- **State / data fetching:** `@tanstack/react-query` 5.90
- **Styling:** Tailwind CSS 4.2 via `@tailwindcss/vite` (CSS-first v4 model; config lives in `frontend/src/index.css`)
- **UI primitives:** `@radix-ui/react-slot`, `@radix-ui/react-tooltip`, `class-variance-authority`, `clsx`, `tailwind-merge`, `lucide-react 0.577`, `@tailwindcss/typography`
- **Drag-and-drop:** `@dnd-kit/core` 6.3, `@dnd-kit/modifiers` 9.0, `@dnd-kit/sortable` 10.0, `@dnd-kit/utilities` 3.2
- **Markdown:** `react-markdown` 10.1, `remark-gfm` 4.0
- **Dev tools:** ESLint 9.39, Prettier 3.8, Vitest 4.0 (`happy-dom` environment), Playwright 1.58

### Infrastructure

`docker-compose.yml` defines three services:

| Service | Container | Host port | Container port | Notes |
|---|---|---|---|---|
| `backend` | `ghchat-backend` | 8000 | 8000 | FastAPI / Uvicorn |
| `frontend` | `ghchat-frontend` | 5173 | 8080 | nginx static server |
| `signal-api` | `ghchat-signal-api` | internal only | 8080 | `bbernhard/signal-cli-rest-api` |

- Backend health: `GET http://localhost:8000/api/v1/health`
- Frontend health: `GET http://localhost:8080/health` (inside container); `http://localhost:5173` from host
- Data volume: `ghchat-data` mounted at `/var/lib/ghchat/data` (SQLite database)
- Signal config volume: `signal-cli-config` at `/home/.local/share/signal-cli`
- All three services share the `ghchat-network` bridge network

## Architecture Notes

- **Auth:** GitHub OAuth with secure HTTP-only session cookies. No JWT / `python-jose` layer.
- **Real-time:** Native WebSocket (`ConnectionManager` in `services/websocket.py`), with SSE fallback in the projects API. Do not reintroduce `socket.io-client`.
- **Storage:** SQLite via `aiosqlite` in WAL mode. Migrations (`001`–`020`) run automatically on startup from `backend/src/migrations/`. Key tables: `user_sessions`, `project_settings`, `agent_configs`, `pipeline_configs`, `mcp_configurations`, `agent_tool_associations`, `chat_messages`, `chat_proposals`, `chat_recommendations`, `chores`, `blocking_queue`.
- **Tailwind v4:** CSS-first config lives entirely in `frontend/src/index.css`. `tailwind.config.js` and `postcss.config.js` are intentionally absent — do not recreate them.
- **Repository resolution:** Use the shared `resolve_repository()` helper (`src/utils.py`) everywhere. Do not introduce ad-hoc owner/repo fallback logic.
- **AI providers:** `completion_providers.py` abstracts GitHub Copilot SDK (default, user OAuth token) and Azure OpenAI (static keys, optional). Selected via `AI_PROVIDER` env var.
- **Agent pipelines:** Configured in SQLite (`pipeline_configs`) and executed by `services/copilot_polling/` + `services/workflow_orchestrator/`. Copilot polling auto-restarts on container restart using the most-recent persisted session (or `GITHUB_WEBHOOK_TOKEN` fallback).
- **Blocking queue:** `services/blocking_queue.py` implements serial issue activation with per-repo `asyncio.Lock` to prevent double-activation race conditions.
- **Chores:** `services/chores/` manages scheduled recurring tasks (scheduler, counter, chat, template builder).
- **Signal messaging:** `signal_bridge.py` / `signal_chat.py` / `signal_delivery.py` integrate with the Signal sidecar for inbound/outbound AI chat over Signal.
- **MCP tools:** `services/mcp_store.py` + `api/mcp.py` manage MCP server configurations and agent tool associations. `services/tools/presets.py` provides a static catalog of built-in MCP presets (GitHub, Azure, Sentry, Cloudflare, Azure DevOps, Context7, Code Graph Context) served via `GET /api/v1/tools/presets`. `services/tools/service.py` handles per-project tool CRUD and repo MCP config sync.
- **MCP presets flow:** User selects preset on Tools page → draft form → saves as user tool in DB → agent dispatch calls `_resolve_agent_tool_selection()` → `generate_config_files()` writes `mcp-servers:` into `.github/agents/{slug}.agent.md` YAML frontmatter → GitHub reads agent file on assignment.
- **Remote MCP config:** `.github/agents/mcp.json` defines MCP servers available to remote GitHub Custom Agents (e.g., Context7 HTTP endpoint). This file is co-located with agent definitions and read by GitHub.com during coding agent sessions.
- **Encryption:** Fernet (`cryptography` package) used for token-at-rest encryption when `ENCRYPTION_KEY` is set.
- **`AsyncGenerator` typing:** Always include both type parameters for Python 3.12 compatibility: `AsyncGenerator[str, None]`.

## Repo Layout

```text
backend/
  src/
    api/              FastAPI route handlers
                      (agents, auth, board, chat, chores, cleanup, health,
                       mcp, metadata, pipelines, projects, settings, signal,
                       tasks, tools, webhooks, workflow)
    middleware/       Request middleware (request_id context var)
    migrations/       SQL schema migrations (001–020, run on startup)
    models/           Pydantic request/response models
    prompts/          AI prompt templates (issue_generation, task_generation)
    services/         Business logic
      agents/         Agent config CRUD
      chores/         Scheduled chores (scheduler, counter, chat, template)
      copilot_polling/ Copilot PR polling loop and agent output parsing
      tools/          MCP tool service (presets catalog, per-project CRUD, repo sync)
      github_projects/ GitHub Projects v2 GraphQL + REST
      housekeeping/   Session/DB cleanup
      pipelines/      Pipeline config service
      workflow_orchestrator/ Issue workflow state machine
  tests/
    unit/
    integration/
    helpers/

frontend/
  src/
    components/       UI components by domain
                      (agents, auth, board, chat, chores, common,
                       pipeline, settings, tools, ui)
    hooks/            React hooks (useAuth, useChat, usePipelineConfig,
                      useBlockingQueue, useProjects, useChores, etc.)
    layout/           Shell components (AppLayout, AuthGate)
    lib/              Shared utilities (cn, etc.)
    pages/            Route-level pages
                      (AppPage, AgentsPage, AgentsPipelinePage, ChoresPage,
                       ProjectsPage, SettingsPage, ToolsPage)
    services/         HTTP client (api.ts)
    types/            Shared TypeScript types
    utils/            Pure utility helpers
  e2e/                Playwright end-to-end tests
```

## Commands

```bash
# Backend
cd backend && source .venv/bin/activate
ruff check src/ tests/          # lint
ruff format src/ tests/         # format (or --check)
pyright src/                    # type-check
pytest tests/unit/ -q           # fast unit tests
pytest tests/ --cov=src         # full suite with coverage

# Frontend
cd frontend
npm run lint                    # ESLint
npm run type-check              # tsc --noEmit
npm run test                    # Vitest (run once)
npm run build                   # production build
npx playwright test             # E2E
```

## Conventions

- Python: Ruff-driven, 100-character line limit, `known-first-party = ["src"]`.
- TypeScript: strict mode, `@/` path alias maps to `frontend/src`.
- Commits: conventional-commit style — `feat:`, `fix:`, `docs:`, `refactor:`, `test:`, `chore:`.
- Prefer focused, minimal fixes over broad refactors unless the task explicitly calls for architectural work.
- Do **not** recreate deleted compatibility files: `frontend/vite.config.js`, `frontend/tailwind.config.js`, `frontend/postcss.config.js`.
- Agent `.agent.md` files live in `.github/agents/`; corresponding `.prompt.md` shortcuts live in `.github/prompts/`.
- `.github/agents/mcp.json` declares MCP servers for remote GitHub Custom Agents (currently Context7). Do not confuse with `.vscode/mcp.json` (local IDE MCP servers).

## Validation Expectations

- **Backend changes:** validate with `ruff check`, `ruff format --check`, `pyright`, and relevant `pytest` coverage.
- **Frontend changes:** validate with `npm run lint`, `npm run type-check`, `npm run test`, and `npm run build`.
- **Pre-commit hook** (`scripts/pre-commit`): runs ruff format (auto-fix) + ruff lint (auto-fix) + pyright on staged Python files; ESLint (auto-fix) on staged frontend files.
- **Pre-push hook** (`scripts/setup-hooks.sh`): full backend + frontend test gates.
- **CI** (`.github/workflows/ci.yml`): backend uses Python 3.12; frontend uses Node 20. Three jobs: `backend`, `frontend`, `docs` (markdownlint + link-check).
- A known flaky failure can occur in `frontend/src/hooks/useAuth.test.tsx` under full parallel runs — confirm isolated behavior before changing unrelated code.

## Frontend Pattern Notes
- Celestial theme animations and gradients are implemented via shared utility classes in `frontend/src/index.css` (for example, orbiting particles, glow effects, and parallax layers). Reuse these utilities instead of defining component-local `@keyframes` or duplicating animation logic.
- When introducing new UI libraries or visual patterns, prefer documenting reusable utilities and conventions here (and in component-level docs) rather than adding branch-specific or version-specific notes.
## Active Technologies
- TypeScript ~5.9 (frontend-only feature) + React 19.2, Tailwind CSS v4.2, class-variance-authority 0.7, lucide-react 0.577 (031-chat-helper-text)
- N/A — placeholder text is static; no database or localStorage changes (031-chat-helper-text)
- TypeScript ~5.9 (frontend), Python 3.13 (backend) + React 19.2, TanStack Query v5.90, Tailwind CSS v4, lucide-react 0.577 (frontend); FastAPI 0.135, aiosqlite 0.22, Pydantic v2.12 (backend) (032-issue-upload-pipeline-config)
- SQLite with WAL mode (aiosqlite) — existing `pipeline_configs` table, existing `project_settings` table (`assigned_pipeline_id` column) (032-issue-upload-pipeline-config)
- TypeScript ~5.9 (frontend), Python 3.13 (backend) + React 19.2, TanStack Query v5.90, Tailwind CSS v4 (with `@tailwindcss/vite`), shadcn/ui (Radix UI primitives), dnd-kit (drag-and-drop), lucide-react 0.577 (icons), class-variance-authority 0.7.1 (frontend); FastAPI 0.135, aiosqlite 0.22, Pydantic v2.12 (backend) (033-audit-pipelines-ux)
- SQLite with WAL mode (aiosqlite) — `pipeline_configs` table for pipeline CRUD (033-audit-pipelines-ux)

## Custom Agents

All agents live in `.github/agents/`. The repository includes both **Spec Kit pipeline agents** and **utility agents**:

### Spec Kit Pipeline Agents
| Agent | Purpose |
|-------|---------|
| `speckit.specify` | Feature specification from issue description |
| `speckit.plan` | Implementation plan with research and data model |
| `speckit.tasks` | Actionable task list from spec + plan |
| `speckit.implement` | Code implementation from tasks |
| `speckit.clarify` | Identify underspecified areas in a spec |
| `speckit.analyze` | Cross-artifact consistency analysis |
| `speckit.checklist` | Custom checklist generation |
| `speckit.constitution` | Project constitution management |
| `speckit.taskstoissues` | Convert tasks to GitHub issues |

### Utility Agents
| Agent | Purpose |
|-------|---------|
| `archivist` | Updates documentation and README to match code changes |
| `designer` | Creates or refines design assets scoped to changes |
| `judge` | Triages PR review comments and applies justified changes |
| `linter` | Runs linting, tests, CI steps, and resolves errors |
| `quality-assurance` | Scoped quality improvements and defect fixes |
| `tester` | Adds tests for changed behavior and improves testability |

### MCP Configuration
- `.github/agents/mcp.json` — Declares MCP servers available to remote GitHub Custom Agents (currently Context7 for documentation lookup).

## MCP Presets

The Tools page exposes a **Preset Library** of built-in MCP server configurations. Presets are defined statically in `backend/src/services/tools/presets.py` and served via `GET /api/v1/tools/presets`.

| Preset | Type | Category | Description |
|--------|------|----------|-------------|
| GitHub MCP Server | HTTP | GitHub | Read-only GitHub MCP server |
| GitHub MCP Server (Full Access) | HTTP | GitHub | Full-access GitHub MCP server |
| Azure MCP | Local | Cloud | Azure-aware coding workflows |
| Sentry MCP | Local | Monitoring | Issue details and summaries |
| Cloudflare MCP | SSE | Cloud | Cloudflare docs and platform access |
| Azure DevOps MCP | Local | Cloud | Azure DevOps work items |
| Context7 | HTTP | Documentation | Up-to-date library docs and code examples |
| Code Graph Context | Local | Code Analysis | Code indexing, call chains, dead code detection |

## Recent Changes
- 033-audit-pipelines-ux: Added TypeScript ~5.9 (frontend), Python 3.13 (backend) + React 19.2, TanStack Query v5.90, Tailwind CSS v4 (with `@tailwindcss/vite`), shadcn/ui (Radix UI primitives), dnd-kit (drag-and-drop), lucide-react 0.577 (icons), class-variance-authority 0.7.1 (frontend); FastAPI 0.135, aiosqlite 0.22, Pydantic v2.12 (backend)
